### PR TITLE
Fix TensorCopy bug

### PR DESCRIPTION
**Pull request by chengduoZH**
Fix tensorcopy bug

### CODE REVIEW

File: paddle/fluid/framework/tensor_util.cc [hunk 1]
```diff
@@ -69,19 +69,47 @@ void TensorCopy(const Tensor& src, const platform::Place& dst_place,
...
+      // be carried out correctly, we should make ctx wait.
+      // If ctx_place and src_place are the same, we should add ctx.Wait()
+      // after memory::Copy; if ctx_place and dst_place are the same, we should
+      // add ctx.Wait() before memory::Copy.
+      if (platform::is_same_place(ctx_place, src_place)) {
+        memory::Copy(dst_gpu_place, dst_ptr, src_gpu_place, src_ptr, size,
+                     stream);
+        ctx.Wait();
+      } else if (platform::is_same_place(ctx_place, dst_place)) {
+        ctx.Wait();
```
Comment (panyx0718) [inline paddle/fluid/framework/tensor_util.cc]: Here should wait on src_ctx

File: paddle/fluid/framework/tensor_util.cc [hunk 2]
```diff
@@ -69,19 +69,47 @@ void TensorCopy(const Tensor& src, const platform::Place& dst_place,
...
 
 void TensorCopy(const Tensor& src, const platform::Place& dst_place,
                 Tensor* dst) {
+  // NOTE(zcd): If the src.place() and dst_place are two different GPU,
+  // the copy operation is carried out on the dst_place's stream. This is
+  // very important, because TensorCopy is an async operator, and in most
+  // case, once this copy operator returns, dst is to be used in dst_place's
+  // stream, if this copy operation is carried out on the src_place's stream,
+  // when dst is used in dst_place's stream the copy operation may be
+  // not completed.
```
Comment (panyx0718) [inline paddle/fluid/framework/tensor_util.cc]: put the doc about sync and async behavior in header.

File: paddle/fluid/framework/tensor_util.cc (hunk 1 above)
Comment (chengduoZH) [reply paddle/fluid/framework/tensor_util.cc]: Done

File: paddle/fluid/framework/tensor_util.cc (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/framework/tensor_util.cc]: Done

**Pull request closed without merging**